core-rust/state-manager/src/state_manager.rs

```diff
@@ -650,28 +650,36 @@ where
         let commit_request_start_state_version =
             commit_state_version.relative(-(commit_transactions_len as i128));
 
-        // Whilst we could validate intent hash duplicates here, these are checked by validators on prepare already,
-        // and the check will move into the engine at some point and we'll get it for free then...
-        let prepare_results = commit_request
-            .transactions
-            .iter()
-            .map(|raw_transaction| {
-                self.ledger_transaction_validator
-                    .prepare_from_raw(raw_transaction)
-            })
-            .collect::<Result<Vec<_>, _>>();
-
-        let prepared_transactions = match prepare_results {
-            Ok(prepared_transactions) => prepared_transactions,
-            Err(error) => {
-                warn!(
-                    "invalid commit request: cannot parse transaction to be committed: {:?}",
-                    error
-                );
-                return Err(InvalidCommitRequestError::TransactionParsingFailed);
+        // Step 1.: Parse the transactions (and collect specific metrics from them, as a drive-by)
+        let mut prepared_transactions = Vec::new();
+        let mut leader_round_counters_builder = LeaderRoundCountersBuilder::default();
+        for (index, raw_transaction) in commit_request.transactions.iter().enumerate() {
+            let result = self
+                .ledger_transaction_validator
+                .prepare_from_raw(raw_transaction);
+            let prepared_transaction = match result {
+                Ok(prepared_transaction) => prepared_transaction,
+                Err(error) => {
+                    warn!(
+                        "invalid commit request: cannot parse transaction at index {}: {:?}",
+                        index, error
+                    );
+                    return Err(InvalidCommitRequestError::TransactionParsingFailed);
+                }
+            };
+
+            if let PreparedLedgerTransactionInner::RoundUpdateV1(_) = &prepared_transaction.inner {
```

Comment:
_(note from author)_

Yes, this is ugly, since we have to re-parse the transaction that was already parsed. Is there a way to use the already-parsed one? Or to capture the `RoundUpdateV1` representation _before_ it becomes `instructions: Vec<_>`?
(From performance PoV, I am sure we can afford this - it only happens for the single round update transaction, not every transaction)

There are also some completely different approaches possible (e.g. emit this information in an Event from Engine - but parsing an Event is not the easiest either!)